scooby/report.py

```diff
@@ -68,118 +59,82 @@ def date(self):
-                       additional=None, sort=False):
-        self._packages = {} # Holds name of packages and their version
-        self._failures = {} # Holds failures and reason
-        self._sort = sort
+    """Internal helper class to access Python info and package versions."""
 
-        # Make sure arguments are good
-        def safety(x):
-            if isinstance(x, (str, ModuleType)):
-                return [x,]
-            elif x is None or len(x) < 1:
-                return []
-            return list(x)
+    def __init__(self, additional, core, optional, sort):
+        self._packages = {}  # Holds name of packages and their version
+        self._sort = sort
 
-        core = safety(core)
-        optional = safety(optional)
-        additional = safety(additional)
+        # Add packages in the following order:
+        self.add_packages(additional)               # Provided by the user
+        self.add_packages(core)                     # Provided by a module dev
+        self.add_packages(optional, optional=True)  # Optional packages
 
-        # Always show scooby as the very last package reported
-        additional.append('scooby')
+    def add_packages(self, packages, optional=False):
+        """Add all packages to list; optional ones only if available."""
 
-        # First listed packages
-        self.add_packages(core)
-        # Optional packages to appear after the core
-        self.add_packages(optional, optional=True)
-        # Additional packages: these are user specified
-        self.add_packages(additional)
+        # Ensure arguments are a list
+        if isinstance(packages, (str, ModuleType)):
+            pckgs = [packages, ]
+        elif packages is None or len(packages) < 1:
+            pckgs = list()
+        else:
+            pckgs = list(packages)
 
+        # Loop over packages
+        for pckg in pckgs:
+            self.get_version(pckg, optional)
 
-    def _safe_import_by_name(self, name, optional=False):
-        try:
-            module = importlib.import_module(name)
-        except ImportError:
-            if not optional:
-                self._failures[name] = UNAVAILABLE_MSG
-            module = None
-        return module
+    def get_version(self, pckg, optional):
+        """Get the version of a package by passing the package or it's name"""
 
+        if isinstance(pckg, str):  # Case 1: pckg is a string; import it
+            name = pckg
+            pckg = self._safe_import_by_name(pckg)
 
-    def _add_package(self, module, name=None, optional=False):
-        """Internal helper to update the packages dictionary with a module
-        """
-        if name is None or not isinstance(name, str):
-            name = module.__name__
-        if not isinstance(module, ModuleType):
-            if optional:
+            # Return if we cannot load the module and it is an optional one
+            if pckg is None and optional:
                 return
```

Comment:
Should this return a "not available string"?

Reply:
I know the never use it if its optional but someone else might on the report object

Reply:
Not sure I understand you right. If you return a `'not available'` here, it would go back to `add_packages`, where it would be lost. We could return it there to, so it would go back to `__init__`, where it gets currently lost too. So we would have to return it there too. But then `Report` would only return a string if there is a package that cannot be found and is optional. Does that make sense?

Or am I missing something. Let's phrase it otherwise: What would be the use-case?

Reply:
The point is: `get_version` never returns anything, it just adds 
```
# Add the version to the package reference
self._packages[name] = version
```
So I don't see why it should return something if an optional package was not found?

Reply:
> Let's phrase it otherwise: What would be the use-case?

Actually.. I don't think there is a use case. If it's optional and not available then I think returning `None` like this is fine - if it weren't optional then it returns the message for not available which is implemented so I think this is good.

Reply:
Exactly. But it wouldn't return it, it would put `self._packages[name] = 'Could not import'`